awsrds/rds_db_instance.go

```diff
@@ -603,15 +624,16 @@ func (r *RDSDBInstance) GetFullValidTargetVersion(engine string, currentVersion
 		targetVersions = append(targetVersions, *target.EngineVersion)
 	}
 
-	targetMonikerSemVer, err := semver.NewVersion(targetVersionMoniker)
-	if err != nil {
-		logSess.Error("parse-target-version-moniker-as-semver", err)
-		return "", err
-	}
-
 	semVersions, err := parseSemanticVersions(targetVersions)
 	targettableVersions := filterTargetVersion(semVersions, engine, *targetMonikerSemVer)
 	if len(targettableVersions) == 0 {
+		if currentSemVersion.Major() == targetMonikerSemVer.Major() {
+			logSess.Info("no-new-version-but-same-major-noop", lager.Data{
+				"target-version-moniker": targetVersionMoniker,
+				"current-version": currentVersion,
+			})
+			return "", nil
+		}
```

Comment:
May be a stupid question, but why do we do this check for "no upgrade targets" twice? If we omit the first time we do this, surely it will fall through and be caught here anyway...

The only difference I can see is we get a little more precision in the error message, but I'm not sure that's worth the extra weirdness.

Reply:
So there are two situations:

1. We are at the latest version available and there are no upgrades available (e.g. latest postgres 14 version)
2. We are at the latest upgrade available for this major version.

Both these situations are handled differently in the current code. Do you want me to refactor?

Reply:
Situation 1 seems like a subset of situation 2 to me. Unless I'm missing something in both of those cases we return `"", nil`, and if we omit the first check it will surely be caught by the second. No?

Reply:
You are correct. The additional if statement here:

https://github.com/alphagov/paas-rds-broker/blob/main/awsrds/rds_db_instance.go#L595

probably isn't needed. I'd need to go through all the logic afterwards and test to ensure this though.

I was trying to minimise the risk and changes when I initially started this.

Want me to refactor and try to remove the initial check?

Reply:
I bet if you rm that clause you won't even need to change the tests, unless you're sniffing for log messages.

Reply:
(maybe if you're checking calls to mocks etc some adjustment will be needed)

Reply:
I'll see if I can get away with just removing it.

Reply:
removed